mesonbuild/build.py

```diff
@@ -320,16 +303,52 @@ def process_sourcelist(self, sources):
                 msg = 'Bad source of type {!r} in target {!r}.'.format(type(s).__name__, self.name)
                 raise InvalidArguments(msg)
 
+    @staticmethod
+    def can_compile_remove_sources(compiler, sources):
+        removed = False
+        for s in sources:
+            if compiler.can_compile(s):
+                sources.remove(s)
```

Comment:
You shouldn't modify a list on which you are iterating; things get skipped in the iteration, and I assume you don't want to skip the one after a match.